### PR TITLE
Share user sessions through the database

### LINKED ISSUE #1145: Make user login session handling not depend on a particular server instance

### PR DESCRIPTION
**Pull request by whisperity**
> Closes #1145.

Along with the existing storage in the server's memory, user authentication session tokens will also be saved to the *configuration* database, in a new table. This way, if multiple servers are used, the user can authenticate through either and still receive the same token as they had before.

When a session cookie is presented, the following steps are taken to verify it:
 * Check if the token exists in the server's memory. If so, previous procedures apply.
 * Check if the token exists in the database. If the user is authenticated properly and the token exists, we create a **local** session by reusing the token in the database.

Session expiry times in the configuration file apply to how tokens *in the database* also expire. Expired tokens are not reused, at the attempt of reusal, they are deleted from the DB too.

Explicit logout also removes tokens from database.

### CODE REVIEW

**Comment by csordasmarton** [comment]
@whisperity The migration script has been merged. Please resolve the conflicts to review you pull request.

File: libcodechecker/server/database/config_db_model.py [hunk 1]
```diff
@@ -106,6 +107,22 @@ def __init__(self, permission, product_id, name, is_group=False):
         self.name = name
         self.is_group = is_group
 
+
+class Session(Base):
```
Comment (gyorb) [inline libcodechecker/server/database/config_db_model.py]: Wouldn't we mix the current database sessions with the user login sessions? I'm just thinking about a different naming here.

File: libcodechecker/libclient/product_helper.py [hunk 2]
```diff
@@ -30,8 +31,7 @@ def __init__(self, protocol, host, port, uri, session_token=None):
         self.client = codeCheckerProductService.Client(self.protocol)
 
         if session_token:
-            headers = {'Cookie': session_manager.SESSION_COOKIE_NAME +
-                       "=" + session_token}
+            headers = {'Cookie': SESSION_COOKIE_NAME + '=' + session_token}
```
Comment (csordasmarton) [inline libcodechecker/libclient/product_helper.py]: You use this line of code in multiple places. Can we create a helper function for it?

File: libcodechecker/server/session_manager.py [hunk 3]
```diff
@@ -0,0 +1,541 @@
+# -------------------------------------------------------------------------
```
Comment (csordasmarton) [inline libcodechecker/server/session_manager.py]: Did you modified this file or just changed the path? Why github doesn't recognize that the path was changed?

File: libcodechecker/server/database/config_db_model.py (hunk 1 above)
Comment (whisperity) [reply libcodechecker/server/database/config_db_model.py]: You import them with different names using `from ... import ... as ...`. The user login session, as an object, is an internal thing to `session_manager.py`, and should not be imported by code outside of it.

File: libcodechecker/server/session_manager.py [hunk 4]
```diff
@@ -0,0 +1,541 @@
...
+
+            for session in self.__sessions[:]:
+                if session.token == token:
+                    self.__sessions.remove(session)
+
+                    if transaction:
+                        records = transaction.query(SessionRecord). \
+                            filter(and_(SessionRecord.auth_string ==
+                                        session.persistent_hash,
+                                        SessionRecord.token == token)).all()
```
Comment (csordasmarton) [inline libcodechecker/server/session_manager.py]: Can't we just call the `delete` operation on the query?
```py
.query().filter().delete()
```

File: libcodechecker/libclient/product_helper.py (hunk 2 above)
Comment (whisperity) [reply libcodechecker/libclient/product_helper.py]: I don't think there exists a good place where this helper function could be put.

File: libcodechecker/server/session_manager.py (hunk 3 above)
Comment (whisperity) [reply libcodechecker/server/session_manager.py]: The first commit split the file into a serverside and a clientside code. The second commit modified the serverside file.

File: config/version.json [hunk 5]
```diff
@@ -6,7 +6,7 @@
     },
     "product_db_version": {
         "major" : "6",
-        "minor" : "0"
+        "minor" : "1"
```
[the remark this answers is not in the document]
Comment (whisperity) [reply config/version.json]: I'm adding this file back but without changing the version inside. Removing these dicts tumbles down into edits required that should not be part of this or any of my other PRs. See #1204.

File: libcodechecker/server/server.py [hunk 6]
```diff
@@ -884,16 +886,42 @@ def start_server(config_directory, package_data, port, db_conn_string,
...
+                os.rename(session_cfg_file, server_cfg_file)
+            else:
+                LOG.info("CodeChecker server's example configuration file "
+                         "created at " + server_cfg_file)
+                shutil.copyfile(example_cfg_file, server_cfg_file)
+
     try:
-        manager = session_manager.SessionManager(root_sha, force_auth)
+        manager = session_manager.SessionManager(
+            server_cfg_file,
```
Comment (csordasmarton) [inline libcodechecker/server/server.py]: I am getting the following error when starting CodeChecker with an empty workspace:
```
UnboundLocalError: local variable 'server_cfg_file' referenced before assignment
```

Comment (whisperity) [reply libcodechecker/server/server.py]: This is fixed. Damn you IDE for pasting code after an `else:` like it was meant to be in the else's statement body.

File: libcodechecker/server/session_manager.py [hunk 7]
```diff
@@ -4,38 +4,38 @@
...
 import os
 import shutil
-import stat
 import time
 import uuid
 
-import portalocker
-
 from libcodechecker.logger import get_logger
+<<<<<<< HEAD:libcodechecker/session_manager.py
```
Comment (gyorb) [inline libcodechecker/server/session_manager.py]: Could you remove these git marks from the source?

File: libcodechecker/server/session_manager.py [hunk 8]
```diff
@@ -444,6 +412,7 @@ def __cleanup_sessions(self):
                                            in SessionManager.__valid_sessions
                                            if s.still_reusable()]
         self.__logins_since_prune = 0
+<<<<<<< HEAD:libcodechecker/session_manager.py
```
Comment (gyorb) [inline libcodechecker/server/session_manager.py]: git marks here too.

File: libcodechecker/server/session_manager.py (hunk 7 above)
Comment (whisperity) [reply libcodechecker/server/session_manager.py]: Done these too.

**Comment by gyorb** [comment]
@whisperity do we have a better place to share information between the server and the client? The name of `version.py` does not reflect this.

**Comment by whisperity** [comment]
@gyorb We don't. I was thinking on extending the API to tell the client what is the cookie's name, but then the command line client would need to be rewritten to store for each individual server not only the token but the cookie name too, essentially slowly turning the CLI into a terminal webbrowser...

These are sort of global constants that need to be shared between the faces in a particular build.

I have an idea for making the build script generate the relevant files that contain the constants, kinda like how configure generates a header used in an autotools project. Could do this as a later patch.

**Pull request merged**